File: BE_Book_Food/src/main/java/com/example/ecommerce/book_food/service/CategoryService.java
package com.example.ecommerce.book_food.service;

import com.example.ecommerce.book_food.dto.request.CreateCategoryRequest;
import com.example.ecommerce.book_food.dto.request.UpdateCategoryRequest;
import com.example.ecommerce.book_food.dto.respone.CategoryResponse;
import com.example.ecommerce.book_food.entity.Category;
import com.example.ecommerce.book_food.exception.CategoryAlreadyExistsException;
import com.example.ecommerce.book_food.exception.CategoryHasFoodsException;
import com.example.ecommerce.book_food.exception.CategoryNotFoundException;
import com.example.ecommerce.book_food.mapper.CategoryMapper;
import com.example.ecommerce.book_food.repository.CategoryRepository;
import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@Transactional
@RequiredArgsConstructor
public class CategoryService {

    private final CategoryRepository categoryRepository;
    private final CategoryMapper categoryMapper;

    //Tạo danh mục mới.
    public CategoryResponse createCategory(CreateCategoryRequest request) throws CategoryAlreadyExistsException {
        if(categoryRepository.existsByNameIgnoreCase(request.getName())) {
            throw new CategoryAlreadyExistsException("Category name is already: " + request.getName());
        }
        Category category = Category.builder()
                .name(request.getName())
                .description(request.getDescription())
                .build();
        Category savedCategory = categoryRepository.save(category);
        return categoryMapper.toRespone(savedCategory);
    }

    //Cập nhật danh mục.
    public CategoryResponse updateCategory(Long id, UpdateCategoryRequest request) throws CategoryNotFoundException, CategoryAlreadyExistsException {
        Category category = categoryRepository.findById(id)
                .orElseThrow(() -> new CategoryNotFoundException("Category not found"));

        if (categoryRepository.existsByNameIgnoreCase(request.getName())
                && !category.getName().equalsIgnoreCase(request.getName())) {
            throw new CategoryAlreadyExistsException("Category name already exists: " + request.getName());
        }

        category.setName(request.getName());
        category.setDescription(request.getDescription());

        Category updateCategory = categoryRepository.save(category);
        return categoryMapper.toRespone(updateCategory);
    }

    //Lấy danh sách tất cả danh mục.
    public Page<CategoryResponse> getAllCategories(int page, int size) {
        Pageable pageable = PageRequest.of(page, size, Sort.by("id").ascending()); // sắp xếp theo id tăng dần
        Page<Category> categoryPage = categoryRepository.findAll(pageable);
        return categoryPage.map(categoryMapper::toRespone);
    }

    //Lấy danh mục theo id
    public CategoryResponse getCategoryById(Long id) {
        Category category = categoryRepository.findById(id)
                .orElseThrow(() -> new CategoryNotFoundException("Category not found with id: " + id));
        return categoryMapper.toRespone(category);
    }

    //Xóa danh mục (nếu không chứa món ăn).
    public void deleteCategory(Long id) throws CategoryHasFoodsException {
        Category category = categoryRepository.findById(id)
                .orElseThrow(() -> new CategoryNotFoundException("Category not found with id: " + id));

        //kiểm tra xem category này có chứa món ăn không
        if(category.getFoods() != null && !category.getFoods().isEmpty()) {
            throw new CategoryHasFoodsException("Cannot delete category because it contains foods");
        }
        categoryRepository.delete(category);

    }

    //đếm tổng số category
    public long countCategories(){
        return categoryRepository.countCategories();
    }

}

File: fe_book_food/src/services/categoryService.ts
import { CategoryResponse, CreateCategoryRequest, UpdateCategoryRequest } from "../types/category";
import { Page } from "../types/page";
import { API_ENDPOINTS, TOKEN_KEY } from "../utils/constants";
import { api, ApiResponse } from "./api";

export const categoryService = {
    // lấy tất cả categories có phân trang
    async getAllCategories(page = 0, size = 10): Promise<Page<CategoryResponse>> {
        const response = await api.get<ApiResponse<Page<CategoryResponse>>>(
            API_ENDPOINTS.CATEGORIES.ALL, {params : {page, size}}
        );
        return response.data.data;
    },

    // lấy danh mục theo id
    async getCategoryById(id: number): Promise<CategoryResponse> {
        const response = await api.get<ApiResponse<CategoryResponse>>(API_ENDPOINTS.CATEGORIES.BY_ID(id));
        return response.data.data;
    },

    // tạo mới category (ADMIN)
    async createCategory(newCategory: CreateCategoryRequest): Promise<CategoryResponse> {
        const response = await api.post<ApiResponse<CategoryResponse>>(API_ENDPOINTS.CATEGORIES.CREATE, newCategory);
        return response.data.data;
    },

    // cập nhật category (ADMIN)
    async updateCategory(id: number, updateCategory: UpdateCategoryRequest): Promise<CategoryResponse> {

        const response = await api.put<ApiResponse<CategoryResponse>>( API_ENDPOINTS.CATEGORIES.UPDATE(id), updateCategory);
        return response.data.data;
    },

    // xóa category (ADMIN)
    async deleteCategory(id: number): Promise<void> {
        const response = await api.delete<ApiResponse<string>>(API_ENDPOINTS.CATEGORIES.DELETE(id));
    },

    // lấy tổng số lượng categories
    async countCategories(): Promise<number> {
        const response = await api.get<ApiResponse<number>>(API_ENDPOINTS.CATEGORIES.TOTAL);
        return response.data.data;
    },

    // tìm kiếm categories theo tên
    async searchCategories(keywords: string): Promise<CategoryResponse[]> {
        const response = await api.get<ApiResponse<CategoryResponse[]>>(
            API_ENDPOINTS.CATEGORIES.SEARCH, {params : {keywords}}
        );
        return response.data.data;
    },

}